In [1]:
%load_ext sql
%sql postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
%sql SELECT song_name FROM billboard_top_100_year_end WHERE year = 1988 AND year_rank = 2


 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
1 rows affected.


[('Need You Tonight',)]

In [2]:
%sql Select * FROM coaches

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
91 rows affected.


[(0, 'AGHAKOUCHEKI Abbas', 'Islamic Republic of Iran', 'IRI', 'Wheelchair Basketball', 'WBK', 'Men'),
 (1, 'AKIN Ahmet', 'Turkey', 'TUR', 'Wheelchair Basketball', 'WBK', 'Men'),
 (2, 'AKPWEH Toussaint', 'France', 'FRA', 'Football 5-a-side', 'FB5', 'Men'),
 (3, 'ARIMA Masato', 'Japan', 'JPN', 'Wheelchair Basketball', 'WBK', None),
 (4, 'BAN Nicole', 'Canada', 'CAN', 'Sitting Volleyball', 'VBS', 'Women'),
 (5, 'BARGUEIRAS MARTINEZ Jesus David', 'Spain', 'ESP', 'Football 5-a-side', 'FB5', 'Men'),
 (6, 'BHANIA Haj', 'Great Britain', 'GBR', 'Wheelchair Basketball', 'WBK', None),
 (7, 'BOUTIBA Saad', 'Algeria', 'ALG', 'Goalball', 'GBL', 'Men'),
 (8, 'BRAHIMI Mustapha', 'Algeria', 'ALG', 'Wheelchair Basketball', 'WBK', 'Men'),
 (9, 'CARRION RUIZ Abraham', 'Spain', 'ESP', 'Wheelchair Basketball', 'WBK', 'Women'),
 (10, 'CHEN Qi', "People's Republic of China", 'CHN', 'Wheelchair Basketball', 'WBK', 'Women'),
 (11, 'CHOUDHRY Gaz', 'Great Britain', 'GBR', 'Wheelchair Basketball', 'WBK', 'Men'),
 (12, 'CORR Peter', 'Australia', 'AUS', 'Goalball', 'GBL', 'Women'),
 (13, 'COTE Patrick', 'Canada', 'CAN', 'Wheelchair Rugby', 'WRU', 'Mixed'),
 (14, 'CUSIN Olivier', 'France', 'FRA', 'Wheelchair Rugby', 'WRU', 'Mixed'),
 (15, 'CZECHOWSKI Jake', 'United States of America', 'USA', 'Goalball', 'GBL', 'Women'),
 (16, 'de RICK Johan', 'Belgium', 'BEL', 'Goalball', 'GBL', 'Men'),
 (17, 'DEMONTE Martin', 'Argentina', 'ARG', 'Football 5-a-side', 'FB5', 'Men'),
 (18, 'DO NASCIMENTO CARDOSO Ana Lucia', 'Colombia', 'COL', 'Wheelchair Basketball', 'WBK', 'Men'),
 (19, 'DUBBERLEY Brad', 'Australia', 'AUS', 'Wheelchair Rugby', 'WRU', 'Mixed'),
 (20, 'DUBROVIN Anton', 'Ukraine', 'UKR', 'Goalball', 'GBL', 'Men'),
 (21, 'DUBROVIN Fedir', 'Ukraine', 'UKR', 'Goalball', 'GBL', 'Men'),
 (22, 'DUCHARME Marc Antoine', 'Canada', 'CAN', 'Wheelchair Basketball', 'WBK', 'Women'),
 (23, 'EDRIS Gharib', 'Egypt', 'EGY', 'Sitting Volleyball', 'VBS', None),
 (24, 'EL MOUNTAQI Driss', 'Morocco', 'MAR', 'Football 5-a-side', 'FB5', 'Men'),
 (25, 'ELAIUTY Mosaad', 'Egypt', 'EGY', 'Sitting Volleyball', 'VBS', 'Men'),
 (26, 'ELIA GIMENO Raul', 'Spain', 'ESP', 'Wheelchair Basketball', 'WBK', None),
 (27, 'ELTAYEB Hani', 'Egypt', 'EGY', 'Goalball', 'GBL', 'Women'),
 (28, 'FAREBROTHER Trent', 'Canada', 'CAN', 'Goalball', 'GBL', 'Women'),
 (29, 'FERIANI Matteo', 'Canada', 'CAN', 'Wheelchair Basketball', 'WBK', 'Men'),
 (30, 'FISHER Simon', 'Great Britain', 'GBR', 'Wheelchair Basketball', 'WBK', 'Women'),
 (31, 'FREITAS DO NASCIMENTO Dailton', 'Brazil', 'BRA', 'Goalball', 'GBL', 'Women'),
 (32, 'FRIDAY Craig', 'Australia', 'AUS', 'Wheelchair Basketball', 'WBK', 'Men'),
 (33, 'FRIDAY Craig', 'Australia', 'AUS', 'Wheelchair Basketball', 'WBK', 'Women'),
 (34, 'FURNESS Shane', 'Australia', 'AUS', 'Wheelchair Basketball', 'WBK', None),
 (35, 'GECEVICIUS Valdas', 'Lithuania', 'LTU', 'Goalball', 'GBL', 'Men'),
 (36, 'GUEDES DANTAS Jose', 'Brazil', 'BRA', 'Sitting Volleyball', 'VBS', 'Women'),
 (37, 'GUENTHER Johannes', 'Germany', 'GER', 'Goalball', 'GBL', 'Men'),
 (38, 'GUMBERT James', 'United States of America', 'USA', 'Wheelchair Rugby', 'WRU', 'Mixed'),
 (39, 'HAMITER William', 'United States of America', 'USA', 'Sitting Volleyball', 'VBS', 'Women'),
 (40, 'HRUSTEMOVIC Mirza', 'Bosnia and Herzegovina', 'BIH', 'Sitting Volleyball', 'VBS', 'Men'),
 (41, 'ICHIKAWA Kyoichi', 'Japan', 'JPN', 'Goalball', 'GBL', 'Women'),
 (42, 'IWASA Yoshiaki', 'Japan', 'JPN', 'Wheelchair Basketball', 'WBK', 'Women'),
 (43, 'JOHNSON Lawrence', 'United States of America', 'USA', 'Wheelchair Basketball', 'WBK', 'Women'),
 (44, 'KARASU Gultekin', 'Turkey', 'TUR', 'Goalball', 'GBL', 'Women'),
 (45, 'KIM Hyun Woo', 'Republic of Korea', 'KOR', 'Wheelchair Basketball', 'WBK', None),
 (46, 'KO Kwangyub', 'Republic of Korea', 'KOR', 'Wheelchair Basketball', 'WBK', None),
 (47, 'KO Kwangyub', 'Republic of Korea', 'KOR', 'Wheelchair Basketball', 'WBK', 'Men'),
 (48, 'KONGDANPRAI Kongkiat', 'Thailand', 'THA', 'Football 5-a-s

In [3]:
%sql SELECT * FROM billboard_top_100_year_end LIMIT 15

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
15 rows affected.


[(0, 1956, 1, 'Elvis Presley', 'Elvis Presley', 'Heartbreak Hotel', 1.0),
 (1, 1956, 2, 'Elvis Presley', 'Elvis Presley', "Don't Be Cruel", 2.0),
 (2, 1956, 3, 'Nelson Riddle', 'Nelson Riddle', 'Lisbon Antigua', 3.0),
 (3, 1956, 4, 'Platters', 'Platters', 'My Prayer', 4.0),
 (4, 1956, 5, 'Gogi Grant', 'Gogi Grant', 'The Wayward Wind', 5.0),
 (5, 1956, 6, 'Les Baxter', 'Les Baxter', 'The Poor People Of Paris', 6.0),
 (6, 1956, 7, 'Doris Day', 'Doris Day', 'Whatever Will Be Will Be (Que Sera Sera)', 7.0),
 (7, 1956, 8, 'Elvis Presley', 'Elvis Presley', 'Hound Dog', 8.0),
 (8, 1956, 9, 'Dean Martin', 'Dean Martin', 'Memories Are Made Of This', 9.0),
 (9, 1956, 10, 'Kay Starr', 'Kay Starr', 'Rock And Roll Waltz', 10.0),
 (10, 1956, 11, 'Morris Stoloff', 'Morris Stoloff', 'Moonglow And Theme From "Picnic"', 11.0),
 (11, 1956, 12, 'Platters', 'Platters', 'The Great Pretender', 12.0),
 (12, 1956, 13, 'Pat Boone', 'Pat Boone', 'I Almost Lost My Mind', 13.0),
 (13, 1956, 14, 'Elvis Presley', 'Elvis Presley', 'I Want You, I Need You, I Love You', 14.0),
 (14, 1956, 15, 'Elvis Presley', 'Elvis Presley', 'Love Me Tender', 15.0)]

In [4]:
%sql SELECT * FROM billboard_top_100_year_end WHERE artist = 'Mariah Carey' LIMIT 10

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
10 rows affected.


[(3502, 1990, 6, 'Mariah Carey', 'Mariah Carey', 'Vision Of Love', 3503.0),
 (3574, 1990, 76, 'Mariah Carey', 'Mariah Carey', 'Love Takes Time', 3575.0),
 (3611, 1991, 13, 'Mariah Carey', 'Mariah Carey', 'Someday', 3612.0),
 (3621, 1991, 22, 'Mariah Carey', 'Mariah Carey', 'Emotions', 3622.0),
 (3625, 1991, 26, 'Mariah Carey', 'Mariah Carey', "I Don't Wanna Cry", 3626.0),
 (3668, 1991, 69, 'Mariah Carey', 'Mariah Carey', 'Love Takes Time', 3669.0),
 (3715, 1992, 16, 'Mariah Carey', 'Mariah Carey', "I'll Be There", 3716.0),
 (3723, 1992, 23, 'Mariah Carey', 'Mariah Carey', "Can't Let Go", 3724.0),
 (3745, 1992, 42, 'Mariah Carey', 'Mariah Carey', 'Make It Happen', 3746.0),
 (3811, 1993, 8, 'Mariah Carey', 'Mariah Carey', 'Dreamlover', 3812.0)]

In [5]:
%sql SELECT year, song_name FROM billboard_top_100_year_end WHERE year_rank = 1 LIMIT 10

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
10 rows affected.


[(1956, 'Heartbreak Hotel'),
 (1957, 'All Shook Up'),
 (1958, 'Volare (Nel Blu Dipinto Di Blu)'),
 (1959, 'The Battle Of New Orleans'),
 (1960, 'Theme From "A Summer Place"'),
 (1961, "Tossin' And Turnin'"),
 (1962, 'Stranger On The Shore'),
 (1963, 'Sugar Shack'),
 (1964, 'I Want To Hold Your Hand'),
 (1965, 'Wooly Bully')]

In [6]:
%%sql

SELECT *
  FROM billboard_top_100_year_end
  ORDER BY artist
  LIMIT 50

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
50 rows affected.


[(4561, 1999, 100, "Alabama feat. 'N Sync", "'N Sync", 'God Must Have Spent A Little More Time On You', 4562.0),
 (4590, 2000, 27, "'N Sync", "'N Sync", "It's Gonna Be Me", 4591.0),
 (4584, 2000, 21, "'N Sync", "'N Sync", 'Bye Bye Bye', 4585.0),
 (4490, 1999, 45, "'N Sync", "'N Sync", '(God Must Have Spent) A Little More Time On You', 4491.0),
 (4365, 1998, 37, "'N Sync", "'N Sync", 'I Want You Back', 4366.0),
 (4848, 2002, 33, "'N Sync feat. Nelly", "'N Sync", 'Girlfriend', 4849.0),
 (4909, 2002, 75, "'N Sync", "'N Sync", 'Gone', 4910.0),
 (4740, 2001, 51, "'N Sync", "'N Sync", 'This I Promise You', 4741.0),
 (4555, 1999, 97, "'N Sync and Gloria Estefan", "'N Sync", 'Music Of My Heart', 4556.0),
 (3046, 1985, 68, "'Til Tuesday", "'Til Tuesday", 'Voices Carry', 3047.0),
 (3457, 1989, 63, '.38 Special', '.38 Special', 'Second Chances', 3458.0),
 (3951, 1994, 40, '10,000 Maniacs', '10,000 Maniacs', 'Because The Night', 3952.0),
 (1500, 1970, 84, '100 Proof Aged In Soul', '100 Proof Aged In Soul', "Somebody's Been Sleeping", 1501.0),
 (1967, 1975, 42, '10cc', '10cc', "I'm Not In Love", 1968.0),
 (2177, 1977, 44, '10cc', '10cc', 'The Things We Do For Love', 2178.0),
 (4149, 1996, 34, '112 feat. The Notorious B.I.G.', '112', 'Only You', 4150.0),
 (4492, 1999, 47, "112 feat. Lil' Zane", '112', 'Anywhere', 4493.0),
 (4268, 1997, 43, 'Allure feat. 112', '112', 'All Cried Out', 4269.0),
 (4260, 1997, 37, '112', '112', 'Cupid', 4261.0),
 (4701, 2001, 20, '112', '112', 'Peaches and Cream', 4702.0),
 (4419, 1998, 84, 'Allure feat. 112', '112', 'All Cried Out', 4420.0),
 (4760, 2001, 68, '112', '112', "It's Over Now", 4761.0),
 (5349, 2005, 97, '112 feat. Foxy Brown', '112', 'U Already Know', 5350.0),
 (1246, 1968, 33, '1910 Fruitgum Company', '1910 Fruitgum Company', 'Simon Says', 1247.0),
 (1261, 1968, 48, '1910 Fruitgum Company', '1910 Fruitgum Company', '1, 2, 3, Red Light', 1262.0),
 (1383, 1969, 67, '1910 Fruitgum Company', '1910 Fruitgum Company', 'Indian Giver', 1384.0),
 (6474, 2013, 99, '2 Chainz', '2 Chainz', "I'm Different", 6475.0),
 (6396, 2013, 41, 'A$AP Rocky feat. Drake, 2 Chainz and Kendrick Lamar', '2 Chainz', 'F**kin Problems', 6397.0),
 (6239, 2012, 27, 'Kanye West, Big Sean, Pusha T, 2 Chainz', '2 Chainz', 'Mercy', 6240.0),
 (6278, 2012, 54, '2 Chainz feat. Drake', '2 Chainz', 'No Lie', 6279.0),
 (5775, 2008, 100, '2 Pistols feat. T-Pain and Tay Dizm', '2 Pistols', 'She Got It', 5776.0),
 (4089, 1995, 76, '20 Fingers feat. Gillette', '20 Fingers', 'Short Dick Man', 4090.0),
 (5059, 2003, 80, '2Pac', '2Pac', 'Thugz Mansion', 5060.0),
 (4063, 1995, 51, '2Pac', '2Pac', 'Dear Mama / Old School', 4064.0),
 (4425, 1998, 89, '2Pac', '2Pac', 'Do For Love', 4426.0),
 (3862, 1993, 56, '2Pac', '2Pac', 'I Get Around', 3863.0),
 (3973, 1994, 61, '2Pac', '2Pac', 'Keep Ya Head Up', 3974.0),
 (4132, 1996, 17, '2Pac', '2Pac', 'How Do U Want It / California Love', 4133.0),
 (5102, 2004, 15, '3 Doors Down', '3 Doors Down', 'Here Without You', 5103.0),
 (4732, 2001, 44, '3 Doors Down', '3 Doors Down', 'Kryptonite', 4733.0),
 (4943, 2003, 5, '3 Doors Down', '3 Doors Down', "When I'm Gone", 4944.0),
 (5704, 2008, 52, '3 Doors Down', '3 Doors Down', "It'S Not My Time", 5705.0),
 (5032, 2003, 62, '3 Doors Down', '3 Doors Down', 'Here Without You', 5033.0),
 (4578, 2000, 15, '3 Doors Down', '3 Doors Down', 'Kryptonite', 4579.0),
 (4792, 2001, 96, '3 Doors Down', '3 Doors Down', 'Be Like That', 4793.0),
 (5275, 2005, 38, '3 Doors Down', '3 Doors Down', 'Let Me Go', 5276.0),
 (2711, 1982, 53, '38 Special', '38 Special', 'Caught Up In You', 2712.0),
 (4761, 2001, 69, '3LW', '3LW', "No More (Baby I'ma Do Right)", 4762.0),
 (5815, 2009, 30, '3OH!3', '3OH!3', "Don't Trust Me", 5816.0),
 (5981, 2010, 51, 'Ke$ha feat. 3OH!3', '3OH!3', 'Blah Blah Blah', 5982.0)]

In [7]:
%%sql

SELECT *
  FROM billboard_top_100_year_end
  ORDER BY artist, year_rank DESC
  LIMIT 50

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
50 rows affected.


[(4561, 1999, 100, "Alabama feat. 'N Sync", "'N Sync", 'God Must Have Spent A Little More Time On You', 4562.0),
 (4555, 1999, 97, "'N Sync and Gloria Estefan", "'N Sync", 'Music Of My Heart', 4556.0),
 (4909, 2002, 75, "'N Sync", "'N Sync", 'Gone', 4910.0),
 (4740, 2001, 51, "'N Sync", "'N Sync", 'This I Promise You', 4741.0),
 (4490, 1999, 45, "'N Sync", "'N Sync", '(God Must Have Spent) A Little More Time On You', 4491.0),
 (4365, 1998, 37, "'N Sync", "'N Sync", 'I Want You Back', 4366.0),
 (4848, 2002, 33, "'N Sync feat. Nelly", "'N Sync", 'Girlfriend', 4849.0),
 (4590, 2000, 27, "'N Sync", "'N Sync", "It's Gonna Be Me", 4591.0),
 (4584, 2000, 21, "'N Sync", "'N Sync", 'Bye Bye Bye', 4585.0),
 (3046, 1985, 68, "'Til Tuesday", "'Til Tuesday", 'Voices Carry', 3047.0),
 (3457, 1989, 63, '.38 Special', '.38 Special', 'Second Chances', 3458.0),
 (3951, 1994, 40, '10,000 Maniacs', '10,000 Maniacs', 'Because The Night', 3952.0),
 (1500, 1970, 84, '100 Proof Aged In Soul', '100 Proof Aged In Soul', "Somebody's Been Sleeping", 1501.0),
 (2177, 1977, 44, '10cc', '10cc', 'The Things We Do For Love', 2178.0),
 (1967, 1975, 42, '10cc', '10cc', "I'm Not In Love", 1968.0),
 (5349, 2005, 97, '112 feat. Foxy Brown', '112', 'U Already Know', 5350.0),
 (4419, 1998, 84, 'Allure feat. 112', '112', 'All Cried Out', 4420.0),
 (4760, 2001, 68, '112', '112', "It's Over Now", 4761.0),
 (4492, 1999, 47, "112 feat. Lil' Zane", '112', 'Anywhere', 4493.0),
 (4268, 1997, 43, 'Allure feat. 112', '112', 'All Cried Out', 4269.0),
 (4260, 1997, 37, '112', '112', 'Cupid', 4261.0),
 (4149, 1996, 34, '112 feat. The Notorious B.I.G.', '112', 'Only You', 4150.0),
 (4701, 2001, 20, '112', '112', 'Peaches and Cream', 4702.0),
 (1383, 1969, 67, '1910 Fruitgum Company', '1910 Fruitgum Company', 'Indian Giver', 1384.0),
 (1261, 1968, 48, '1910 Fruitgum Company', '1910 Fruitgum Company', '1, 2, 3, Red Light', 1262.0),
 (1246, 1968, 33, '1910 Fruitgum Company', '1910 Fruitgum Company', 'Simon Says', 1247.0),
 (6474, 2013, 99, '2 Chainz', '2 Chainz', "I'm Different", 6475.0),
 (6278, 2012, 54, '2 Chainz feat. Drake', '2 Chainz', 'No Lie', 6279.0),
 (6396, 2013, 41, 'A$AP Rocky feat. Drake, 2 Chainz and Kendrick Lamar', '2 Chainz', 'F**kin Problems', 6397.0),
 (6239, 2012, 27, 'Kanye West, Big Sean, Pusha T, 2 Chainz', '2 Chainz', 'Mercy', 6240.0),
 (5775, 2008, 100, '2 Pistols feat. T-Pain and Tay Dizm', '2 Pistols', 'She Got It', 5776.0),
 (4089, 1995, 76, '20 Fingers feat. Gillette', '20 Fingers', 'Short Dick Man', 4090.0),
 (4425, 1998, 89, '2Pac', '2Pac', 'Do For Love', 4426.0),
 (5059, 2003, 80, '2Pac', '2Pac', 'Thugz Mansion', 5060.0),
 (3973, 1994, 61, '2Pac', '2Pac', 'Keep Ya Head Up', 3974.0),
 (3862, 1993, 56, '2Pac', '2Pac', 'I Get Around', 3863.0),
 (4063, 1995, 51, '2Pac', '2Pac', 'Dear Mama / Old School', 4064.0),
 (4132, 1996, 17, '2Pac', '2Pac', 'How Do U Want It / California Love', 4133.0),
 (4792, 2001, 96, '3 Doors Down', '3 Doors Down', 'Be Like That', 4793.0),
 (5032, 2003, 62, '3 Doors Down', '3 Doors Down', 'Here Without You', 5033.0),
 (5704, 2008, 52, '3 Doors Down', '3 Doors Down', "It'S Not My Time", 5705.0),
 (4732, 2001, 44, '3 Doors Down', '3 Doors Down', 'Kryptonite', 4733.0),
 (5275, 2005, 38, '3 Doors Down', '3 Doors Down', 'Let Me Go', 5276.0),
 (4578, 2000, 15, '3 Doors Down', '3 Doors Down', 'Kryptonite', 4579.0),
 (5102, 2004, 15, '3 Doors Down', '3 Doors Down', 'Here Without You', 5103.0),
 (4943, 2003, 5, '3 Doors Down', '3 Doors Down', "When I'm Gone", 4944.0),
 (2711, 1982, 53, '38 Special', '38 Special', 'Caught Up In You', 2712.0),
 (4761, 2001, 69, '3LW', '3LW', "No More (Baby I'ma Do Right)", 4762.0),
 (6015, 2010, 73, '3OH!3 feat. Ke$ha', '3OH!3', 'My First Kiss', 6016.0),
 (5981, 2010, 51, 'Ke$ha feat. 3OH!3', '3OH!3', 'Blah Blah Blah', 5982.0)]

In [8]:
%%sql

SELECT * 
  FROM billboard_top_100_year_end 
  WHERE artist = 'Aaliyah' 
  ORDER BY year_rank

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
11 rows affected.


[(4946, 2003, 8, 'Aaliyah', 'Aaliyah', 'Miss You', 4947.0),
 (4575, 2000, 12, 'Aaliyah', 'Aaliyah', 'Try Again', 4576.0),
 (3935, 1994, 24, 'Aaliyah', 'Aaliyah', 'Back and Forth', 3936.0),
 (4884, 2002, 58, 'Aaliyah', 'Aaliyah', 'More than a Woman', 4885.0),
 (4188, 1996, 69, 'Aaliyah', 'Aaliyah', 'If Your Girl Only Knew', 4189.0),
 (3982, 1994, 70, 'Aaliyah', 'Aaliyah', 'At Your Best (You Are Love)', 3983.0),
 (4305, 1997, 78, 'Aaliyah', 'Aaliyah', 'The One I Gave My Heart To', 4306.0),
 (4921, 2002, 85, 'Aaliyah', 'Aaliyah', 'Rock the Boat', 4922.0),
 (4422, 1998, 86, 'Aaliyah', 'Aaliyah', 'The One I Gave My Heart To', 4423.0),
 (5071, 2003, 92, 'Aaliyah', 'Aaliyah', 'Come Over', 5072.0),
 (4671, 2000, 96, 'Aaliyah', 'Aaliyah', "I Don't Wanna", 4672.0)]

In [9]:
%%sql

SELECT * FROM billboard_top_100_year_end 
  WHERE year_rank = 1
  ORDER BY artist

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
64 rows affected.


[(4938, 2003, 1, '50 Cent', '50 Cent', 'In Da Club', 4939.0),
 (3907, 1994, 1, 'Ace Of Base', 'Ace Of Base', 'The Sign', 3908.0),
 (6055, 2011, 1, 'Adele', 'Adele', 'Rolling In The Deep', 6056.0),
 (2234, 1978, 1, 'Andy Gibb', 'Andy Gibb', 'Shadow Dancing', 2235.0),
 (1317, 1969, 1, 'Archies', 'Archies', 'Sugar, Sugar', 1318.0),
 (3183, 1987, 1, 'Bangles', 'Bangles', 'Walk Like An Egyptian', 3184.0),
 (1822, 1974, 1, 'Barbra Streisand', 'Barbra Streisand', 'The Way We Were', 1823.0),
 (809, 1964, 1, 'Beatles', 'Beatles', 'I Want To Hold Your Hand', 810.0),
 (1214, 1968, 1, 'Beatles', 'Beatles', 'Hey Jude', 1215.0),
 (5494, 2007, 1, 'Beyonce', 'Beyonce', 'Irreplaceable', 5495.0),
 (2441, 1980, 1, 'Blondie', 'Blondie', 'Call Me', 2442.0),
 (508, 1961, 1, 'Bobby Lewis', 'Bobby Lewis', "Tossin' And Turnin'", 509.0),
 (3700, 1992, 1, 'Boyz II Men', 'Boyz II Men', 'End Of The Road', 3701.0),
 (3599, 1991, 1, 'Bryan Adams', 'Bryan Adams', '(Everything I Do) I Do It For You', 3600.0),
 (1926, 1975, 1, 'Captain and Tennille', 'Captain and Tennille', 'Love Will Keep Us Together', 1927.0),
 (4438, 1999, 1, 'Cher', 'Cher', 'Believe', 4439.0),
 (3395, 1989, 1, 'Chicago', 'Chicago', 'Look Away', 3396.0),
 (4013, 1995, 1, 'Coolio', 'Coolio', "Gangsta's Paradise", 4014.0),
 (5355, 2006, 1, 'Daniel Powter', 'Daniel Powter', 'Bad Day', 5356.0),
 (3080, 1986, 1, 'Dionne and Friends', 'Dionne and Friends', "That's What Friends Are For", 3081.0),
 (203, 1958, 1, 'Domenico Modugno', 'Domenico Modugno', 'Volare (Nel Blu Dipinto Di Blu)', 204.0),
 (4221, 1997, 1, 'Elton John', 'Elton John', 'Candle In The Wind 1997', 4222.0),
 (103, 1957, 1, 'Elvis Presley', 'Elvis Presley', 'All Shook Up', 104.0),
 (0, 1956, 1, 'Elvis Presley', 'Elvis Presley', 'Heartbreak Hotel', 1.0),
 (4562, 2000, 1, 'Faith Hill', 'Faith Hill', 'Breathe', 4563.0),
 (5637, 2008, 1, 'Flo Rida feat. T-Pain', 'Flo Rida', 'Low', 5638.0),
 (3289, 1988, 1, 'George Michael', 'George Michael', 'Faith', 3290.0),
 (6202, 2012, 1, 'Gotye feat. Kimbra', 'Gotye', 'Somebody That I Used To Know', 6203.0),
 (708, 1963, 1, 'Jimmy Gilmer and The Fireballs', 'Jimmy Gilmer and The Fireballs', 'Sugar Shack', 709.0),
 (304, 1959, 1, 'Johnny Horton', 'Johnny Horton', 'The Battle Of New Orleans', 305.0),
 (5908, 2010, 1, 'Ke$ha', 'Ke$ha', 'TiK ToK', 5909.0),
 (2545, 1981, 1, 'Kim Carnes', 'Kim Carnes', 'Bette Davis Eyes', 2546.0),
 (6203, 2012, 1, 'Gotye feat. Kimbra', 'Kimbra', 'Somebody That I Used To Know', 6204.0),
 (2338, 1979, 1, 'Knack', 'Knack', 'My Sharona', 2339.0),
 (4676, 2001, 1, 'Lifehouse', 'Lifehouse', 'Hanging By A Moment', 4677.0),
 (5082, 2004, 1, 'Usher feat. Lil Jon and Ludacris', 'Lil Jon', 'Yeah!', 5083.0),
 (4115, 1996, 1, 'Los Del Rio', 'Los Del Rio', 'Macarena (Bayside Boys Mix)', 4116.0),
 (5083, 2004, 1, 'Usher feat. Lil Jon and Ludacris', 'Ludacris', 'Yeah!', 5084.0),
 (1111, 1967, 1, 'Lulu', 'Lulu', 'To Sir With Love', 1112.0),
 (6333, 2013, 1, 'Macklemore and Ryan Lewis feat. Wanz', 'Macklemore', 'Thrift Shop', 6334.0),
 (5220, 2005, 1, 'Mariah Carey', 'Mariah Carey', 'We Belong Together', 5221.0),
 (608, 1962, 1, 'Mr. Acker Bilk', 'Mr. Acker Bilk', 'Stranger On The Shore', 609.0),
 (4329, 1998, 1, 'Next', 'Next', 'Too Close', 4330.0),
 (4799, 2002, 1, 'Nickelback', 'Nickelback', 'How You Remind Me', 4800.0),
 (2655, 1982, 1, 'Olivia Newton-John', 'Olivia Newton-John', 'Physical', 2656.0),
 (406, 1960, 1, 'Percy Faith', 'Percy Faith', 'Theme From "A Summer Place"', 407.0),
 (2760, 1983, 1, 'Police', 'Police', 'Every Breath You Take', 2761.0),
 (2869, 1984, 1, 'Prince', 'Prince', 'When Doves Cry', 2870.0),
 (1617, 1972, 1, 'Roberta Flack', 'Roberta Flack', 'The First Time Ever I Saw Your Face', 1618.0),
 (2132, 1977, 1, 'Rod Stewart', 'Rod Stewart', "Tonight's The Night (Gonna Be Alright)", 2133.0),
 (6334, 2013, 1, 'Macklemore and Ryan Lewis feat. Wanz', 'Ryan Lewis', 'Thrift Shop', 6335.0),
 (911, 1965, 1, 'Sam The Sham and The Pharaohs', 'Sam The Sham and The Pharaohs', '

In [10]:
%%sql

SELECT * FROM billboard_top_100_year_end 
  WHERE year = 1979
  ORDER BY year_rank DESC

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
103 rows affected.


[(2440, 1979, 100, 'Firefall', 'Firefall', 'Strange Way', 2441.0),
 (2439, 1979, 99, 'G.Q.', 'G.Q.', 'I Do Love You', 2440.0),
 (2438, 1979, 98, 'Commodores', 'Commodores', 'Sail On', 2439.0),
 (2437, 1979, 97, 'Diana Ross', 'Diana Ross', 'The Boss', 2438.0),
 (2436, 1979, 96, 'Nigel Olsson', 'Nigel Olsson', 'Dancing Shoes', 2437.0),
 (2435, 1979, 95, 'Van Halen', 'Van Halen', 'Dance The Night Away', 2436.0),
 (2434, 1979, 94, 'Bob Seger and The Silver Bullet Band', 'Bob Seger and The Silver Bullet Band', "We've Got Tonight", 2435.0),
 (2433, 1979, 93, 'Barry Manilow', 'Barry Manilow', 'Somewhere In The Night', 2434.0),
 (2432, 1979, 92, 'Robert Palmer', 'Robert Palmer', "Bad Case Of Lovin' You", 2433.0),
 (2431, 1979, 91, 'Michael Jackson', 'Michael Jackson', "Don't Stop 'Til You Get Enough", 2432.0),
 (2430, 1979, 90, 'Instant Funk', 'Instant Funk', 'I Got My Mind Made Up', 2431.0),
 (2429, 1979, 89, 'Babys', 'Babys', 'Every Time I Think Of You', 2430.0),
 (2428, 1979, 88, 'Foreigner', 'Foreigner', 'Double Vision', 2429.0),
 (2427, 1979, 87, 'Alice Cooper', 'Alice Cooper', 'How You Gonna See Me Now', 2428.0),
 (2426, 1979, 86, 'Rex Smith', 'Rex Smith', 'You Take My Breath Away', 2427.0),
 (2425, 1979, 85, 'Eddie Rabbitt', 'Eddie Rabbitt', 'Suspicions', 2426.0),
 (2424, 1979, 84, 'Ambrosia', 'Ambrosia', 'How Much I Feel', 2425.0),
 (2423, 1979, 83, 'Roger Voudouris', 'Roger Voudouris', 'Get Used To It', 2424.0),
 (2422, 1979, 82, 'Eric Clapton', 'Eric Clapton', 'Promises', 2423.0),
 (2421, 1979, 81, 'Electric Light Orchestra', 'Electric Light Orchestra', "Don't Bring Me Down", 2422.0),
 (2420, 1979, 80, 'Herb Alpert', 'Herb Alpert', 'Rise', 2421.0),
 (2419, 1979, 79, 'Al Stewart', 'Al Stewart', 'Time Passages', 2420.0),
 (2418, 1979, 78, 'Earth, Wind and Fire', 'Earth, Wind and Fire', 'September', 2419.0),
 (2417, 1979, 77, 'Linda Ronstadt', 'Linda Ronstadt', 'Ooh Baby Baby', 2418.0),
 (2416, 1979, 76, 'G.Q.', 'G.Q.', 'Disco Nights', 2417.0),
 (2415, 1979, 75, 'Gino Vannelli', 'Gino Vannelli', 'I Just Wanna Stop', 2416.0),
 (2414, 1979, 74, 'Kiss', 'Kiss', "I Was Made For Lovin' You", 2415.0),
 (2413, 1979, 73, 'Ian Matthews', 'Ian Matthews', 'Shake It', 2414.0),
 (2412, 1979, 72, 'Anne Murray', 'Anne Murray', 'I Just Fall In Love Again', 2413.0),
 (2411, 1979, 71, 'Electric Light Orchestra', 'Electric Light Orchestra', 'Shine A Little Love', 2412.0),
 (2410, 1979, 70, 'Patrick Hernandez', 'Patrick Hernandez', 'Born To Be Alive', 2411.0),
 (2409, 1979, 69, 'Cheryl Lynn', 'Cheryl Lynn', 'Got To Be Real', 2410.0),
 (2408, 1979, 68, 'England Dan and John Ford Coley', 'England Dan and John Ford Coley', 'Love Is The Answer', 2409.0),
 (2407, 1979, 67, 'Styx', 'Styx', 'Renegade', 2408.0),
 (2406, 1979, 66, 'Little River Band', 'Little River Band', 'Lonesome Loser', 2407.0),
 (2405, 1979, 65, 'McFadden and Whitehead', 'McFadden and Whitehead', "Ain't No Stoppin' Us Now", 2406.0),
 (2404, 1979, 64, 'Alicia Bridges', 'Alicia Bridges', 'I Love The Night Life', 2405.0),
 (2403, 1979, 63, 'Rickie Lee Jones', 'Rickie Lee Jones', "Chuck E's In Love", 2404.0),
 (2402, 1979, 62, 'Chic', 'Chic', 'I Want Your Love', 2403.0),
 (2401, 1979, 61, 'Dire Straits', 'Dire Straits', 'Sultans Of Swing', 2402.0),
 (2400, 1979, 60, 'Ace Frehley', 'Ace Frehley', 'New York Groove', 2401.0),
 (2399, 1979, 59, 'Bobby Caldwell', 'Bobby Caldwell', "What You Won't Do For Love", 2400.0),
 (2398, 1979, 58, 'Andy Gibb', 'Andy Gibb', "(Our Love) Don't Throw It All Away", 2399.0),
 (2397, 1979, 57, 'Earth, Wind and Fire', 'Earth, Wind and Fire', 'Boogie Wonderland', 2398.0),
 (2396, 1979, 56, 'Cher', 'Cher', 'Take Me Home', 2397.0),
 (2395, 1979, 55, 'Hot Chocolate', 'Hot Chocolate', "Every 1's A Winner", 2396.0),
 (2394, 1979, 54, 'Bad Company', 'Bad Company', "Rock 'N' Roll Fantasy", 2395.0),
 (2393, 1979, 53, 'Sister Sledge', 'Sister Sledge', 'We Are Family', 2394.0),
 (2392, 1979, 52, 'Wings', 'Wings', 'Goodnight Tonight', 2393.0),
 (2391, 1979, 51, 'John Stewart

In [11]:
%%sql 
SELECT * 
FROM billboard_top_100_year_end
WHERE group_name = 'Justin Bieber'

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
2 rows affected.


[(5891, 2009, 89, 'Justin Bieber', 'Justin Bieber', 'One Time', 5892.0),
 (6240, 2012, 28, 'Justin Bieber', 'Justin Bieber', 'Boyfriend', 6241.0)]

In [12]:
%%sql 
SELECT * 
FROM billboard_top_100_year_end
WHERE group_name LIKE '%Justin Bieber%'

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
12 rows affected.


[(5891, 2009, 89, 'Justin Bieber', 'Justin Bieber', 'One Time', 5892.0),
 (5970, 2010, 44, 'Justin Bieber feat. Ludacris', 'Justin Bieber', 'Baby', 5971.0),
 (5971, 2010, 44, 'Justin Bieber feat. Ludacris', 'Ludacris', 'Baby', 5972.0),
 (6041, 2010, 89, 'Sean Kingston and Justin Bieber', 'Sean Kingston', 'Eenie Meenie', 6042.0),
 (6042, 2010, 89, 'Sean Kingston and Justin Bieber', 'Justin Bieber', 'Eenie Meenie', 6043.0),
 (6240, 2012, 28, 'Justin Bieber', 'Justin Bieber', 'Boyfriend', 6241.0),
 (6250, 2012, 34, 'Justin Bieber feat. Big Sean', 'Justin Bieber', 'As Long as You Love Me', 6251.0),
 (6251, 2012, 34, 'Justin Bieber feat. Big Sean', 'Big Sean', 'As Long as You Love Me', 6252.0),
 (6398, 2013, 42, 'Justin Bieber feat. Nicki Minaj', 'Justin Bieber', 'Beauty And A Beat', 6399.0),
 (6399, 2013, 42, 'Justin Bieber feat. Nicki Minaj', 'Nicki Minaj', 'Beauty And A Beat', 6400.0),
 (6469, 2013, 95, 'will.i.am feat. Justin Bieber', 'will.i.am', '#thatPOWER', 6470.0),
 (6470, 2013, 95, 'will.i.am feat. Justin Bieber', 'Justin Bieber', '#thatPOWER', 6471.0)]

In [13]:
%%sql 
SELECT * 
FROM billboard_top_100_year_end
WHERE artist = 'Santana'
AND group_name NOT LIKE '%feat.%'

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
2 rows affected.


[(1485, 1970, 69, 'Santana', 'Santana', 'Evil Ways', 1486.0),
 (2635, 1981, 84, 'Santana', 'Santana', 'Winning', 2636.0)]

In [14]:
%%sql 
SELECT * 
FROM billboard_top_100_year_end
WHERE year_rank = 1
AND year BETWEEN 1990 AND 1999

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
10 rows affected.


[(3497, 1990, 1, 'Wilson Phillips', 'Wilson Phillips', 'Hold On', 3498.0),
 (3599, 1991, 1, 'Bryan Adams', 'Bryan Adams', '(Everything I Do) I Do It For You', 3600.0),
 (3700, 1992, 1, 'Boyz II Men', 'Boyz II Men', 'End Of The Road', 3701.0),
 (3804, 1993, 1, 'Whitney Houston', 'Whitney Houston', 'I Will Always Love You', 3805.0),
 (3907, 1994, 1, 'Ace Of Base', 'Ace Of Base', 'The Sign', 3908.0),
 (4013, 1995, 1, 'Coolio', 'Coolio', "Gangsta's Paradise", 4014.0),
 (4115, 1996, 1, 'Los Del Rio', 'Los Del Rio', 'Macarena (Bayside Boys Mix)', 4116.0),
 (4221, 1997, 1, 'Elton John', 'Elton John', 'Candle In The Wind 1997', 4222.0),
 (4329, 1998, 1, 'Next', 'Next', 'Too Close', 4330.0),
 (4438, 1999, 1, 'Cher', 'Cher', 'Believe', 4439.0)]

In [15]:
%%sql
SELECT * FROM billboard_top_100_year_end
WHERE artist = 'Queen'
AND year BETWEEN 1980 AND 1989

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
2 rows affected.


[(2446, 1980, 6, 'Queen', 'Queen', 'Crazy Little Thing Called Love', 2447.0),
 (2615, 1981, 65, 'Queen', 'Queen', 'Another One Bites The Dust', 2616.0)]

In [16]:
%%sql 
SELECT * 
FROM billboard_top_100_year_end
WHERE artist LIKE 'Taylor%'
AND artist != 'Taylor Swift'

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
9 rows affected.


[(3309, 1988, 20, 'Taylor Dayne', 'Taylor Dayne', "I'll Always Love You", 3310.0),
 (3343, 1988, 53, 'Taylor Dayne', 'Taylor Dayne', 'Tell It To My Heart', 3344.0),
 (3394, 1988, 100, 'Taylor Dayne', 'Taylor Dayne', 'Prove Your Love', 3395.0),
 (3432, 1989, 38, 'Taylor Dayne', 'Taylor Dayne', "Don't Rush Me", 3433.0),
 (3526, 1990, 28, 'Taylor Dayne', 'Taylor Dayne', 'Love Will Lead You Back', 3527.0),
 (3546, 1990, 48, 'Taylor Dayne', 'Taylor Dayne', 'With Every Beat Of My Heart', 3547.0),
 (3561, 1990, 63, 'Taylor Dayne', 'Taylor Dayne', "I'll Be Your Shelter", 3562.0),
 (3905, 1993, 99, 'Taylor Dayne', 'Taylor Dayne', "Can't Get Enough Of Your Love", 3906.0),
 (5492, 2006, 99, 'Taylor Hicks', 'Taylor Hicks', 'Do I Make you Proud', 5493.0)]

Find all of Rihanna's collaborations (those which included another featured artist), which charted outside the top 50. Don't include any Rihanna solo tracks.

In [17]:
%%sql 
SELECT * 
FROM billboard_top_100_year_end
WHERE artist = 'Rihanna'
AND group_name != 'Rihanna'
AND year_rank > 50

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
5 rows affected.


[(5611, 2007, 85, 'Rihanna and Sean Paul', 'Rihanna', 'Break It Off', 5612.0),
 (5718, 2008, 62, 'Rihanna feat. Ne-Yo', 'Rihanna', 'Hate That I Love You', 5719.0),
 (6157, 2011, 69, 'Rihanna feat. Calvin Harris', 'Rihanna', 'We Found Love', 6158.0),
 (6307, 2012, 79, 'Rihanna feat. Chris Brown', 'Rihanna', 'Birthday Cake', 6308.0),
 (6425, 2013, 59, 'Wale feat. Tiara Thomas or Rihanna', 'Rihanna', 'Bad', 6426.0)]

In [18]:
%sql SELECT *, ("High Price" - "Low Price") / 2 + "Low Price" AS "Midpoint Price" FROM us_pumpkin_prices LIMIT 5;

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
5 rows affected.


[(0, 'PUMPKINS', 'ATLANTA', None, '24 inch bins', 'HOWDEN TYPE', None, None, datetime.date(2016, 9, 24), 140.0, 154.75, 140.0, 154.75, 'MICHIGAN', None, 'jbo', None, None, None, None, None, None, None, None, 'N', None, 147.375),
 (1, 'PUMPKINS', 'ATLANTA', None, '24 inch bins', 'HOWDEN TYPE', None, None, datetime.date(2016, 9, 24), 145.0, 154.75, 145.0, 154.75, 'MICHIGAN', None, 'xlge', None, None, None, None, None, None, None, None, 'N', None, 149.875),
 (2, 'PUMPKINS', 'ATLANTA', None, '24 inch bins', 'HOWDEN TYPE', None, None, datetime.date(2016, 9, 24), 150.0, 154.75, 150.0, 154.75, 'MICHIGAN', None, 'med-lge', None, None, None, None, None, None, None, None, 'N', None, 152.375),
 (3, 'PUMPKINS', 'ATLANTA', None, '24 inch bins', 'HOWDEN TYPE', None, None, datetime.date(2016, 9, 24), 150.0, 150.0, 150.0, 150.0, 'MICHIGAN', None, 'sml', None, None, None, None, None, None, None, None, 'N', None, 150.0),
 (4, 'PUMPKINS', 'ATLANTA', None, '24 inch bins', 'HOWDEN TYPE', None, None, datetime.date(2016, 10, 1), 140.0, 154.75, 140.0, 154.75, 'MICHIGAN', None, 'jbo', None, None, None, None, None, None, None, None, 'N', None, 147.375)]

What happens if you now run SELECT * FROM us_pumpkin_prices;? Do you still see the new column?

In [19]:
%sql SELECT * FROM us_pumpkin_prices LIMIT 5

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
5 rows affected.


[(0, 'PUMPKINS', 'ATLANTA', None, '24 inch bins', 'HOWDEN TYPE', None, None, datetime.date(2016, 9, 24), 140.0, 154.75, 140.0, 154.75, 'MICHIGAN', None, 'jbo', None, None, None, None, None, None, None, None, 'N', None),
 (1, 'PUMPKINS', 'ATLANTA', None, '24 inch bins', 'HOWDEN TYPE', None, None, datetime.date(2016, 9, 24), 145.0, 154.75, 145.0, 154.75, 'MICHIGAN', None, 'xlge', None, None, None, None, None, None, None, None, 'N', None),
 (2, 'PUMPKINS', 'ATLANTA', None, '24 inch bins', 'HOWDEN TYPE', None, None, datetime.date(2016, 9, 24), 150.0, 154.75, 150.0, 154.75, 'MICHIGAN', None, 'med-lge', None, None, None, None, None, None, None, None, 'N', None),
 (3, 'PUMPKINS', 'ATLANTA', None, '24 inch bins', 'HOWDEN TYPE', None, None, datetime.date(2016, 9, 24), 150.0, 150.0, 150.0, 150.0, 'MICHIGAN', None, 'sml', None, None, None, None, None, None, None, None, 'N', None),
 (4, 'PUMPKINS', 'ATLANTA', None, '24 inch bins', 'HOWDEN TYPE', None, None, datetime.date(2016, 10, 1), 140.0, 154.75, 140.0, 154.75, 'MICHIGAN', None, 'jbo', None, None, None, None, None, None, None, None, 'N', None)]

The Head of US Pumpkins Limited is back with another question. For each row, what is the difference between High Price and Low Price?

Write a query that will generate a new derived column called Price Range which, for each row, shows the difference between High Price and Low Price.

In [20]:
%%sql 
SELECT *, ("High Price" - "Low Price") 
  AS "Price Range" 
  FROM us_pumpkin_prices LIMIT 5

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
5 rows affected.


[(0, 'PUMPKINS', 'ATLANTA', None, '24 inch bins', 'HOWDEN TYPE', None, None, datetime.date(2016, 9, 24), 140.0, 154.75, 140.0, 154.75, 'MICHIGAN', None, 'jbo', None, None, None, None, None, None, None, None, 'N', None, 14.75),
 (1, 'PUMPKINS', 'ATLANTA', None, '24 inch bins', 'HOWDEN TYPE', None, None, datetime.date(2016, 9, 24), 145.0, 154.75, 145.0, 154.75, 'MICHIGAN', None, 'xlge', None, None, None, None, None, None, None, None, 'N', None, 9.75),
 (2, 'PUMPKINS', 'ATLANTA', None, '24 inch bins', 'HOWDEN TYPE', None, None, datetime.date(2016, 9, 24), 150.0, 154.75, 150.0, 154.75, 'MICHIGAN', None, 'med-lge', None, None, None, None, None, None, None, None, 'N', None, 4.75),
 (3, 'PUMPKINS', 'ATLANTA', None, '24 inch bins', 'HOWDEN TYPE', None, None, datetime.date(2016, 9, 24), 150.0, 150.0, 150.0, 150.0, 'MICHIGAN', None, 'sml', None, None, None, None, None, None, None, None, 'N', None, 0.0),
 (4, 'PUMPKINS', 'ATLANTA', None, '24 inch bins', 'HOWDEN TYPE', None, None, datetime.date(2016, 10, 1), 140.0, 154.75, 140.0, 154.75, 'MICHIGAN', None, 'jbo', None, None, None, None, None, None, None, None, 'N', None, 14.75)]

Find the 5 pumpkins with the smallest Price Range. Yes, this is for the Head of US Pumpkins, so get cracking! These pumpkins will be considered more 'reliable' than the rest, owing to their relatively stable prices.



In [21]:
%%sql 
SELECT *, ("High Price" - "Low Price") 
  AS "Price Range"
  FROM us_pumpkin_prices 
  ORDER BY "Price Range"
  LIMIT 5

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
5 rows affected.


[(11, 'PUMPKINS', 'ATLANTA', None, '24 inch bins', 'HOWDEN TYPE', None, None, datetime.date(2016, 10, 8), 150.0, 150.0, 150.0, 150.0, 'MICHIGAN', None, 'sml', None, None, None, None, None, None, None, None, 'N', None, 0.0),
 (16, 'PUMPKINS', 'ATLANTA', None, '24 inch bins', 'HOWDEN TYPE', None, None, datetime.date(2016, 10, 15), 150.0, 150.0, 150.0, 150.0, 'MICHIGAN', None, 'sml', None, None, None, None, None, None, None, None, 'N', None, 0.0),
 (3, 'PUMPKINS', 'ATLANTA', None, '24 inch bins', 'HOWDEN TYPE', None, None, datetime.date(2016, 9, 24), 150.0, 150.0, 150.0, 150.0, 'MICHIGAN', None, 'sml', None, None, None, None, None, None, None, None, 'N', None, 0.0),
 (7, 'PUMPKINS', 'ATLANTA', None, '24 inch bins', 'HOWDEN TYPE', None, None, datetime.date(2016, 10, 1), 150.0, 150.0, 150.0, 150.0, 'MICHIGAN', None, 'sml', None, None, None, None, None, None, None, None, 'N', None, 0.0),
 (21, 'PUMPKINS', 'ATLANTA', None, '24 inch bins', 'HOWDEN TYPE', None, None, datetime.date(2016, 10, 22), 150.0, 150.0, 150.0, 150.0, 'MICHIGAN', None, 'sml', None, None, None, None, None, None, None, None, 'N', None, 0.0)]

In [22]:
%%sql 
SELECT COUNT(*)
FROM us_pumpkin_prices;

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
1 rows affected.


[(1757,)]

In [23]:
%%sql
SELECT COUNT("Item Size")
FROM us_pumpkin_prices;

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
1 rows affected.


[(1478,)]

In [24]:
%%sql
SELECT SUM("Low Price")
FROM us_pumpkin_prices;

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
1 rows affected.


[(218871.8399999998,)]

In [25]:
%%sql
SELECT MAX("High Price") FROM us_pumpkin_prices;


 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
1 rows affected.


[(480.0,)]

In [26]:
%%sql
SELECT MAX("City Name") FROM us_pumpkin_prices;

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
1 rows affected.


[('ST. LOUIS',)]

In [27]:
%%sql
SELECT AVG("High Price") FROM us_pumpkin_prices;

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
1 rows affected.


[(132.97093910073977,)]

In [28]:
%%sql
SELECT "Variety", COUNT(*) FROM us_pumpkin_prices GROUP BY "Variety";

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
11 rows affected.


[('HOWDEN TYPE', 542),
 ('BLUE TYPE', 19),
 ('BIG MACK TYPE', 74),
 ('MIXED HEIRLOOM VARIETIES', 57),
 (None, 5),
 ('KNUCKLE HEAD', 20),
 ('HOWDEN WHITE TYPE', 49),
 ('CINDERELLA', 81),
 ('PIE TYPE', 468),
 ('MINIATURE', 310),
 ('FAIRYTALE', 132)]

The HAVING keyword is a bit like WHERE. However, WHERE can't be used with aggregate functions, so we can use HAVING instead. For example, we can refine the last query above a little further:

In [29]:
%%sql
SELECT "City Name", "Variety", COUNT(*) 
FROM us_pumpkin_prices 
GROUP BY "City Name", "Variety" 
HAVING MAX("High Price") > 300 
ORDER BY "City Name" ASC;

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
5 rows affected.


[('BALTIMORE', 'CINDERELLA', 25),
 ('COLUMBIA', 'MINIATURE', 54),
 ('LOS ANGELES', 'HOWDEN TYPE', 21),
 ('SAN FRANCISCO', 'PIE TYPE', 45),
 ('SAN FRANCISCO', 'HOWDEN TYPE', 37)]

Write a query that returns the varieties of pumpkins in different cities, grouped by "City Name" that are having a minimum price of 5, and ordered by "Variety".

In [30]:
%%sql
SELECT "City Name", "Variety"
FROM us_pumpkin_prices 
GROUP BY "City Name", "Variety"
HAVING MIN("Low Price") >= 5
ORDER BY "Variety" ASC;

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
75 rows affected.


[('NEW YORK', 'BIG MACK TYPE'),
 ('SAN FRANCISCO', 'BIG MACK TYPE'),
 ('DALLAS', 'BIG MACK TYPE'),
 ('BALTIMORE', 'BIG MACK TYPE'),
 ('BOSTON', 'BIG MACK TYPE'),
 ('BOSTON', 'BLUE TYPE'),
 ('SAN FRANCISCO', 'BLUE TYPE'),
 ('COLUMBIA', 'BLUE TYPE'),
 ('DALLAS', 'CINDERELLA'),
 ('BOSTON', 'CINDERELLA'),
 ('BALTIMORE', 'CINDERELLA'),
 ('COLUMBIA', 'CINDERELLA'),
 ('SAN FRANCISCO', 'CINDERELLA'),
 ('COLUMBIA', 'FAIRYTALE'),
 ('CHICAGO', 'FAIRYTALE'),
 ('BOSTON', 'FAIRYTALE'),
 ('BALTIMORE', 'FAIRYTALE'),
 ('DALLAS', 'FAIRYTALE'),
 ('SAN FRANCISCO', 'FAIRYTALE'),
 ('LOS ANGELES', 'HOWDEN TYPE'),
 ('BOSTON', 'HOWDEN TYPE'),
 ('COLUMBIA', 'HOWDEN TYPE'),
 ('MIAMI', 'HOWDEN TYPE'),
 ('SAN FRANCISCO', 'HOWDEN TYPE'),
 ('DALLAS', 'HOWDEN TYPE'),
 ('CHICAGO', 'HOWDEN TYPE'),
 ('ATLANTA', 'HOWDEN TYPE'),
 ('NEW YORK', 'HOWDEN TYPE'),
 ('ST. LOUIS', 'HOWDEN TYPE'),
 ('BALTIMORE', 'HOWDEN TYPE'),
 ('PHILADELPHIA', 'HOWDEN TYPE'),
 ('DETROIT', 'HOWDEN TYPE'),
 ('DALLAS', 'HOWDEN WHITE TYPE'),
 ('BALTIMORE', 'HOWDEN WHITE TYPE'),
 ('ST. LOUIS', 'HOWDEN WHITE TYPE'),
 ('DETROIT', 'HOWDEN WHITE TYPE'),
 ('CHICAGO', 'HOWDEN WHITE TYPE'),
 ('COLUMBIA', 'HOWDEN WHITE TYPE'),
 ('NEW YORK', 'KNUCKLE HEAD'),
 ('BOSTON', 'KNUCKLE HEAD'),
 ('COLUMBIA', 'KNUCKLE HEAD'),
 ('PHILADELPHIA', 'KNUCKLE HEAD'),
 ('SAN FRANCISCO', 'MINIATURE'),
 ('ST. LOUIS', 'MINIATURE'),
 ('PHILADELPHIA', 'MINIATURE'),
 ('NEW YORK', 'MINIATURE'),
 ('LOS ANGELES', 'MINIATURE'),
 ('BOSTON', 'MINIATURE'),
 ('COLUMBIA', 'MINIATURE'),
 ('ATLANTA', 'MINIATURE'),
 ('CHICAGO', 'MINIATURE'),
 ('BALTIMORE', 'MINIATURE'),
 ('CHICAGO', 'MIXED HEIRLOOM VARIETIES'),
 ('DALLAS', 'MIXED HEIRLOOM VARIETIES'),
 ('COLUMBIA', 'MIXED HEIRLOOM VARIETIES'),
 ('NEW YORK', 'MIXED HEIRLOOM VARIETIES'),
 ('BALTIMORE', 'MIXED HEIRLOOM VARIETIES'),
 ('ST. LOUIS', 'MIXED HEIRLOOM VARIETIES'),
 ('SAN FRANCISCO', 'MIXED HEIRLOOM VARIETIES'),
 ('DETROIT', 'MIXED HEIRLOOM VARIETIES'),
 ('LOS ANGELES', 'PIE TYPE'),
 ('ST. LOUIS', 'PIE TYPE'),
 ('DALLAS', 'PIE TYPE'),
 ('COLUMBIA', 'PIE TYPE'),
 ('MIAMI', 'PIE TYPE'),
 ('SAN FRANCISCO', 'PIE TYPE'),
 ('NEW YORK', 'PIE TYPE'),
 ('BOSTON', 'PIE TYPE'),
 ('BALTIMORE', 'PIE TYPE'),
 ('PHILADELPHIA', 'PIE TYPE'),
 ('CHICAGO', 'PIE TYPE'),
 ('DETROIT', 'PIE TYPE'),
 ('ATLANTA', 'PIE TYPE'),
 ('DALLAS', None),
 ('BALTIMORE', None)]

OK, so how do we count the number of hits each artist has in billboard_top_100_year_end? Write a query that generates the a table where the top 5 rows are...

Artist	Hits

Madonna	36

Elvis Presley	36

Rihanna	33

Mariah Carey	33

Ludacris	28

In [31]:
%%sql
SELECT "artist", COUNT(*) AS "Hits"
FROM billboard_top_100_year_end
GROUP BY "artist"
ORDER BY "Hits" DESC
LIMIT 5

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
5 rows affected.


[('Elvis Presley', 36),
 ('Madonna', 36),
 ('Rihanna', 33),
 ('Mariah Carey', 33),
 ('Ludacris', 28)]

Write a query which returns the difference between the "High Price" and the "Low Price" of pumpkins in different cities. Ensure that the results are returned in descending order.

In [32]:
%%sql 
SELECT "City Name", AVG("High Price" - "Low Price") 
  AS "Difference"
  FROM us_pumpkin_prices 
  GROUP BY "City Name"
  ORDER BY "Difference" DESC

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
13 rows affected.


[('BOSTON', 24.394886363636363),
 ('ATLANTA', 14.232456140350877),
 ('NEW YORK', 9.357142857142858),
 ('BALTIMORE', 7.565359477124183),
 ('LOS ANGELES', 5.812096774193549),
 ('PHILADELPHIA', 5.614035087719298),
 ('SAN FRANCISCO', 5.25),
 ('DALLAS', 4.817518248175182),
 ('CHICAGO', 2.6653225806451615),
 ('DETROIT', 1.2272727272727273),
 ('ST. LOUIS', 1.1407766990291262),
 ('COLUMBIA', 0.41825095057034223),
 ('MIAMI', 0.0)]

Margarida wanted to write a query that returned an ordered list of all athletes who had won more than 1 medal. Her query is below but it doesn't work as intended.

SELECT COUNT(*), name

FROM medals
    
GROUP BY name
    
WHERE count >= 1
    
ORDER BY count DESC;

In [33]:
%%sql 
SELECT COUNT(*), name
    FROM medals
    GROUP BY name
    HAVING COUNT(*) > 1
    ORDER BY count DESC;

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
391 rows affected.


[(7, 'KRYPAK Maksym'),
 (7, 'RAIMONDI Stefano'),
 (6, 'LONG Jessica'),
 (6, 'TRUSOV Andrii'),
 (5, 'BOKI Ihar'),
 (5, 'ZHDANOV Roman'),
 (5, 'FANTIN Antonio'),
 (5, 'RIVARD Aurelie'),
 (5, 'DUNN Reece'),
 (5, 'GOMES SANTIAGO Maria Carolina'),
 (5, 'SUZUKI Takayuki'),
 (5, 'TERZI Giulia'),
 (5, 'SCHAER Manuela'),
 (5, 'WANG Lichao'),
 (5, 'GILLI Carlotta'),
 (5, 'ZIJDERVELD Chantalle'),
 (5, 'MERESHKO Yelyzaveta'),
 (4, 'KRUGER Lisa'),
 (4, 'JIANG Yuyan'),
 (4, 'RONG Jing'),
 (4, 'LAKATOS Brent'),
 (4, 'LEAUTE Alexandre'),
 (4, 'COUTYA Dimitri'),
 (4, 'PASCOE Sophie'),
 (4, 'LU Dong'),
 (4, 'CAI Liwen'),
 (4, 'PIKALOVA Daria'),
 (4, 'BARLAAM Simone'),
 (4, 'PALAZZO Xenia Francesca'),
 (4, 'CRISPIN CORZO Nelson'),
 (4, 'MARTIN William'),
 (4, 'HUG Marcel'),
 (4, 'SHABALINA Valeriia'),
 (4, 'JIA Hongguang'),
 (4, 'FIRTH Bethany'),
 (4, 'ZHENG Tao'),
 (4, 'MAYHUGH Nick'),
 (4, 'TRIMI Arjola'),
 (4, 'BANDEIRA Gabriel'),
 (4, 'SERRANO ZARATE Carlos Daniel'),
 (3, 'ISHCHIULOVA Viktoriia'),
 (3, 'BEGGIATO Luigi'),
 (3, 'METELKA Jozef'),
 (3, 'HU Daoliang'),
 (3, 'CROTHERS Rowan'),
 (3, 'DURAND ELIAS Omara'),
 (3, 'KONKOLY Zsofia'),
 (3, 'VDOVIN Andrei'),
 (3, 'PLAT Jetze'),
 (3, 'HODGE Timothy'),
 (3, 'JIANG Fenfen'),
 (3, 'DUNLEVY Katie-George'),
 (3, 'DORSMAN Rogier'),
 (3, 'GRECO Paige'),
 (3, 'WEGGEMANN Mallory'),
 (3, 'OSTAPCHENKO Denys'),
 (3, 'DADAON Ami Omer'),
 (3, 'de FARIA DIAS Daniel'),
 (3, 'van GASS Jaco'),
 (3, 'BOCCIARDO Francesco'),
 (3, 'VERAKSA Maksym'),
 (3, 'McFADDEN Tatyana'),
 (3, 'de ROZARIO Madison'),
 (3, 'POPHAM Ben'),
 (3, 'VERGNAUD Loic'),
 (3, 'KALINA Andrei'),
 (3, 'APPLEGATE Jessica-Jane'),
 (3, 'HERNANDEZ HERNANDEZ Jesus'),
 (3, 'PEARSON Lee'),
 (3, 'DUBROV Denys'),
 (3, 'CLEGG Stephen'),
 (3, 'LANE-WRIGHT Crystal'),
 (3, 'MASON Brittni'),
 (3, 'LI Zhangyu'),
 (3, 'MALYAR Mark'),
 (3, 'MARKS Elizabeth'),
 (3, 'VERA ANDRADE Lisbeli Marina'),
 (3, 'ZHOU Hongzhuan'),
 (3, 'ABARZA Alberto'),
 (3, 'ZHOU Zhaoqian'),
 (3, 'PAEYO Pongsakorn'),
 (3, 'LOPEZ DIAZ Diego'),
 (3, 'CLIFFORD Jaryd'),
 (3, 'DEMENTYEV Yehor'),
 (3, 'LIU Cuiqing'),
 (3, 'VOETS Sanne'),
 (3, 'BIAN Jing'),
 (3, 'HANCE Benjamin James'),
 (3, 'DANILENKO Vladimir'),
 (3, 'LI Guizhi'),
 (3, 'MALONE Noah'),
 (3, 'SUN Gang'),
 (3, 'SCHOTT Verena'),
 (3, 'GILLIVER Piers'),
 (3, 'HOSMAR Frank'),
 (3, 'NIKOLAEV Andrei'),
 (3, 'ZHANG Li'),
 (3, 'GLOCK Talisson Henrique'),
 (3, 'SALEI Raman'),
 (3, 'YUAN Weiyi'),
 (3, 'BOTURCHUK Oksana'),
 (3, 'TIAN Jianquan'),
 (3, 'TAN Shumei'),
 (3, 'ZHOU Yanfei'),
 (3, 'BOGGIONI Monica'),
 (3, 'JOUANNY Florian'),
 (3, 'BELARMINO PEREIRA Wendell'),
 (3, 'TOMITA Uchu'),
 (3, 'BAKER Natasha'),
 (3, 'WEN Xiaoyan'),
 (3, 'van GANSEWINKEL Marlene'),
 (3, 'MAZZONE Luca'),
 (3, 'FERNANDEZ INFANTE Marta'),
 (3, 'LEANI RATRI Oktila'),
 (3, 'MA Jia'),
 (3, 'PAP Bianka'),
 (3, 'TRUNNELL Roxanne'),
 (3, 'dos SANTOS ARAUJO Gabriel Geraldo'),
 (3, 'MARTIN Raymond'),
 (2, 'SHABUROV Maxim'),
 (2, 'HABSCH Roger'),
 (2, 'SHCHETNIK Iryna'),
 (2, 'KUZYUKOV Alexander'),
 (2, 'PUCH Pepo'),
 (2, 'CHEN Jianxin'),
 (2, 'KIM Young-Gun'),
 (2, 'TARASOV Denis'),
 (2, 'SUN Bianbian'),
 (2, 'van ZON Kelly'),
 (2, 'ZHOU Ying'),
 (2, 'DEBRUNNER Catherine'),
 (2, 'FLOORS Johannes'),
 (2, 'YOON Jiyu'),
 (2, 'YAMAZAKI Yuma'),
 (2, 'ROMANCHUK Daniel'),
 (2, 'FRUEHWIRTH Thomas'),
 (2, 'ALHAMMADI Mohamed'),
 (2, 'CLAEYS Manon'),
 (2, 'DREOCK KAESER Angelika'),
 (2, 'SUGIURA Keiko'),
 (2, 'KALASHIAN Artem'),
 (2, 'ZHAO Lixue'),
 (2, 'DANA Alicia'),
 (2, 'MA Lin'),
 (2, 'SAFRONOV Dmitrii'),
 (2, 'SHISHOVA Iuliia'),
 (2, 'TLILI Raoua'),
 (2, 'PORTAL Alex'),
 (2, 'DAI Yunqiang'),
 (2, 'KIM Jung Jun'),
 (2, 'HICKS Darren'),
 (2, 'IGLESIAS FORNEIRO Adiaratou'),
 (2, 'SUGIMURA Hidetaka'),
 (2, 'VINCENT Isabella'),
 (2, 'SNIKUS Rihards'),
 (2, 'LIU Li'),
 (2, 'CELEN Tim'),
 (2, 'AITCHISON Danielle'),
 (2, 'SHI Yiting'),
 (2, 'PEK Karolina'),
 (2, 'CUNDY Jody'),
 (2, 'CHENG Hefang'),
 (2, 'YAN Shuo'),
 (2, 'ADENEGAN Kare'),
 (2, 'JEBABLI Rouay'),
 (2, 'JANSEN 

Marco wanted to write a query that returns list of countries ordered by the number of medals they won, with the country winning the most medals at the top. Their query is below... it doesn't work as intended.

SELECT

COUNT(*), countrycode

FROM medals

ORDER BY count ASC

GROUP BY country;

In [34]:
%%sql
SELECT
    COUNT(*), country
    FROM medals
    GROUP BY country
    ORDER BY count DESC
    ;

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
86 rows affected.


[(276, "People's Republic of China"),
 (171, 'Great Britain'),
 (167, 'United States of America'),
 (153, 'RPC'),
 (111, 'Brazil'),
 (108, 'Ukraine'),
 (101, 'Australia'),
 (92, 'Japan'),
 (91, 'Italy'),
 (77, 'Netherlands'),
 (69, 'France'),
 (45, 'Germany'),
 (39, 'Spain'),
 (35, 'Islamic Republic of Iran'),
 (33, 'Republic of Korea'),
 (27, 'Poland'),
 (24, 'Thailand'),
 (24, 'Canada'),
 (24, 'Colombia'),
 (22, 'Mexico'),
 (21, 'Turkey'),
 (20, 'Morocco'),
 (19, 'Azerbaijan'),
 (19, 'Uzbekistan'),
 (19, 'India'),
 (18, 'Argentina'),
 (18, 'Hungary'),
 (15, 'Slovakia'),
 (15, 'Belgium'),
 (14, 'Switzerland'),
 (13, 'Greece'),
 (12, 'New Zealand'),
 (12, 'Nigeria'),
 (12, 'Algeria'),
 (11, 'Bosnia and Herzegovina'),
 (11, 'Tunisia'),
 (11, 'Indonesia'),
 (10, 'Czech Republic'),
 (9, 'Israel'),
 (9, 'Sweden'),
 (9, 'Austria'),
 (8, 'Croatia'),
 (8, 'Lithuania'),
 (7, 'Belarus'),
 (7, 'Serbia'),
 (7, 'Venezuela'),
 (7, 'Egypt'),
 (7, 'Hong Kong, China'),
 (7, 'Ireland'),
 (7, 'South Africa'),
 (6, 'Chile'),
 (6, 'Cuba'),
 (5, 'Jordan'),
 (5, 'Finland'),
 (5, 'Latvia'),
 (5, 'Kazakhstan'),
 (5, 'Malaysia'),
 (5, 'Denmark'),
 (4, 'Norway'),
 (3, 'Iraq'),
 (3, 'Ecuador'),
 (3, 'United Arab Emirates'),
 (3, 'Georgia'),
 (2, 'Singapore'),
 (2, 'Costa Rica'),
 (2, 'Cyprus'),
 (2, 'Bulgaria'),
 (2, 'Slovenia'),
 (2, 'Romania'),
 (2, 'Namibia'),
 (2, 'Kuwait'),
 (2, 'Portugal'),
 (2, 'Sri Lanka'),
 (1, 'El Salvador'),
 (1, 'Qatar'),
 (1, 'Pakistan'),
 (1, 'Chinese Taipei'),
 (1, 'Mongolia'),
 (1, 'Kenya'),
 (1, 'Ethiopia'),
 (1, 'Peru'),
 (1, 'Vietnam'),
 (1, 'Saudi Arabia'),
 (1, 'Uganda'),
 (1, 'Montenegro'),
 (1, 'Oman')]

In [35]:
#%sql SELECT * FROM athletes INNER JOIN medals ON athletes.name = medals.name AND athletes.sport = medals.sport;

Answer the original questions that were posed at the start of this bite

Who was the oldest medal winner?

Who was the youngest medal winner?


In [36]:
%%sql
SELECT * FROM athletes 
  INNER JOIN medals 
  ON athletes.name = medals.name 
  AND athletes.sport = medals.sport
  ORDER BY athletes.date_of_birth
  LIMIT 5

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
5 rows affected.


[(1266, 'FREI Heinz', 'Switzerland', 'SUI', 'Cycling Road', 'CRD', datetime.date(1958, 1, 28), 'Male', 923, 'Switzerland', 'SUI', 'FREI Heinz', 'Cycling Road', 'CRD', "Men's H3 Road Race", 'Silver Medal'),
 (3235, 'PUSHPASHEVA Nadezhda', 'RPC', 'RPC', 'Table Tennis', 'TTE', datetime.date(1959, 12, 29), 'Female', 2016, 'RPC', 'RPC', 'PUSHPASHEVA Nadezhda', 'Table Tennis', 'TTE', "Women's Singles - Classes 1-2", 'Bronze Medal'),
 (1153, 'ETZNER JAKOBSSON Louise', 'Sweden', 'SWE', 'Equestrian', 'EQU', datetime.date(1960, 6, 9), 'Female', 984, 'Sweden', 'SWE', 'ETZNER JAKOBSSON Louise', 'Equestrian', 'EQU', 'Individual Freestyle Test - Grade IV', 'Silver Medal'),
 (781, 'COOKE Carol', 'Australia', 'AUS', 'Cycling Road', 'CRD', datetime.date(1961, 8, 6), 'Female', 950, 'Australia', 'AUS', 'COOKE Carol', 'Cycling Road', 'CRD', "Women's T1-2 Time Trial", 'Silver Medal'),
 (4233, 'WALSH Jill', 'United States of America', 'USA', 'Cycling Road', 'CRD', datetime.date(1963, 6, 19), 'Female', 1634, 'United States of America', 'USA', 'WALSH Jill', 'Cycling Road', 'CRD', "Women's T1-2 Road Race", 'Bronze Medal')]

In [37]:
%%sql
SELECT * FROM athletes 
  INNER JOIN medals 
  ON athletes.name = medals.name 
  AND athletes.sport = medals.sport
  ORDER BY athletes.date_of_birth DESC
  LIMIT 5

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
5 rows affected.


[(4346, 'YAMADA Miyuki', 'Japan', 'JPN', 'Swimming', 'SWM', datetime.date(2006, 9, 15), 'Female', 1232, 'Japan', 'JPN', 'YAMADA Miyuki', 'Swimming', 'SWM', "Women's 50m Backstroke - S2", 'Silver Medal'),
 (4346, 'YAMADA Miyuki', 'Japan', 'JPN', 'Swimming', 'SWM', datetime.date(2006, 9, 15), 'Female', 1173, 'Japan', 'JPN', 'YAMADA Miyuki', 'Swimming', 'SWM', "Women's 100m Backstroke - S2", 'Silver Medal'),
 (4187, 'VINCENT Isabella', 'Australia', 'AUS', 'Swimming', 'SWM', datetime.date(2006, 1, 14), 'Female', 1935, 'Australia', 'AUS', 'VINCENT Isabella', 'Swimming', 'SWM', "Women's 4x100m Medley Relay - 34 Points", 'Bronze Medal'),
 (4187, 'VINCENT Isabella', 'Australia', 'AUS', 'Swimming', 'SWM', datetime.date(2006, 1, 14), 'Female', 1227, 'Australia', 'AUS', 'VINCENT Isabella', 'Swimming', 'SWM', "Women's 4x100m Freestyle Relay - 34 Pts", 'Silver Medal'),
 (599, 'CAMACHO RAMIREZ Angel de Jesus', 'Mexico', 'MEX', 'Swimming', 'SWM', datetime.date(2004, 12, 28), 'Male', 1861, 'Mexico', 'MEX', 'CAMACHO RAMIREZ Angel de Jesus', 'Swimming', 'SWM', "Men's 50m Backstroke - S4", 'Bronze Medal')]

Use a SELF JOIN to find athletes that share a birthday.

Bonus challenge: Try to avoid matching an athlete with themselves

In [38]:
# SELECT alias_a.name staff, alias_b.name manager
# FROM staff alias_a INNER JOIN staff alias_b
# ON alias_a.id =  alias_b.manager_id;

In [39]:
%%sql
SELECT COUNT(*)
  FROM athletes 

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
1 rows affected.


[(4527,)]

In [40]:
%%sql
SELECT alias_a.name name, alias_b.name name
FROM athletes alias_a 
INNER JOIN athletes alias_b
ON alias_a.date_of_birth =  alias_b.date_of_birth
WHERE alias_a.name != alias_b.name
ORDER BY alias_a.date_of_birth
LIMIT 50

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
50 rows affected.


[('CHEBANIKA Raisa', 'SIM Jae Yong'),
 ('SIM Jae Yong', 'CHEBANIKA Raisa'),
 ('PILLE-STEPPAT Sylvia', 'SCHNARNDORF Susana'),
 ('SCHNARNDORF Susana', 'PILLE-STEPPAT Sylvia'),
 ('POMME Mauricio', 'FENDT Bernhard'),
 ('FENDT Bernhard', 'POMME Mauricio'),
 ('BACHMAIER Ernst', 'CHIPPINGTON Jeanette'),
 ('CHIPPINGTON Jeanette', 'BACHMAIER Ernst'),
 ('RODRIGUEZ Matthew', 'ANDALOUSSI Ahmed'),
 ('ANDALOUSSI Ahmed', 'RODRIGUEZ Matthew'),
 ('MACEDO Jose', 'YOO Byunghoon'),
 ('YOO Byunghoon', 'MACEDO Jose'),
 ('FUJII Yuriko', 'VENGE BALBOA Christian'),
 ('VENGE BALBOA Christian', 'FUJII Yuriko'),
 ('VENGE BALBOA Christian', 'FUJII Yuriko'),
 ('FUJII Yuriko', 'VENGE BALBOA Christian'),
 ('PETRUNIV Vasyl', 'SHIM Youngjip'),
 ('SHIM Youngjip', 'PETRUNIV Vasyl'),
 ('DETTONI Cristian', 'ABU AWAD Khetam'),
 ('ABU AWAD Khetam', 'DETTONI Cristian'),
 ('PENDER Dariusz', 'di TORO Lisa Daniela'),
 ('di TORO Lisa Daniela', 'PENDER Dariusz'),
 ('SMETANINE David', 'PALIKUCA Mitar'),
 ('PALIKUCA Mitar', 'SMETANINE David'),
 ('WILK Rafal', 'JOHNSON-DWYER Philippa'),
 ('JOHNSON-DWYER Philippa', 'WILK Rafal'),
 ('DAVID Krisztina', 'DIXON Amy'),
 ('DIXON Amy', 'DAVID Krisztina'),
 ('CUNNINGTON Tammy', 'VATNHAMAR Havard'),
 ('VATNHAMAR Havard', 'CUNNINGTON Tammy'),
 ('KIM Kyung Hoon', 'MALYSHEV Sergey'),
 ('MALYSHEV Sergey', 'KIM Kyung Hoon'),
 ('de JESUS SANTOS Edneusa', 'NOVAK Carol-Eduard'),
 ('NOVAK Carol-Eduard', 'de JESUS SANTOS Edneusa'),
 ('de JESUS SANTOS Edneusa', 'NOVAK Carol-Eduard'),
 ('NOVAK Carol-Eduard', 'de JESUS SANTOS Edneusa'),
 ('LIN Ya-Hsuan', 'KORVER Cher'),
 ('KORVER Cher', 'LIN Ya-Hsuan'),
 ('MARVIG Henrik', 'TOUPE David'),
 ('TOUPE David', 'MARVIG Henrik'),
 ('MARVIG Henrik', 'TOUPE David'),
 ('TOUPE David', 'MARVIG Henrik'),
 ('BRUSSIG Carmen', 'BRUSSIG Ramona'),
 ('BRUSSIG Ramona', 'BRUSSIG Carmen'),
 ('JACOBS David', 'EBERHARDT Fernando'),
 ('EBERHARDT Fernando', 'JACOBS David'),
 ('TATSUMI Hiromi', 'JEON Min Jae'),
 ('JEON Min Jae', 'TATSUMI Hiromi'),
 ('de OLIVEIRA Jose Carlos Chagas', 'GUEDOUN Nabil'),
 ('SABINO Claudia Cicero', 'GUEDOUN Nabil')]

In [41]:
%%sql
SELECT DISTINCT alias_a.name AS athlete_1_name, alias_a.date_of_birth AS athlete_1_DOB,  alias_b.name AS athlete_2_name, alias_b.date_of_birth  AS athlete_2_DOB
  FROM athletes alias_a
  INNER JOIN athletes alias_b 
  ON alias_a.date_of_birth = alias_b.date_of_birth 
  WHERE alias_a.name < alias_b.name
  ORDER BY athlete_1_DOB
  LIMIT 100;

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
100 rows affected.


[('CHEBANIKA Raisa', datetime.date(1964, 7, 5), 'SIM Jae Yong', datetime.date(1964, 7, 5)),
 ('PILLE-STEPPAT Sylvia', datetime.date(1967, 10, 12), 'SCHNARNDORF Susana', datetime.date(1967, 10, 12)),
 ('FENDT Bernhard', datetime.date(1970, 3, 24), 'POMME Mauricio', datetime.date(1970, 3, 24)),
 ('BACHMAIER Ernst', datetime.date(1970, 4, 21), 'CHIPPINGTON Jeanette', datetime.date(1970, 4, 21)),
 ('ANDALOUSSI Ahmed', datetime.date(1972, 6, 8), 'RODRIGUEZ Matthew', datetime.date(1972, 6, 8)),
 ('MACEDO Jose', datetime.date(1972, 6, 30), 'YOO Byunghoon', datetime.date(1972, 6, 30)),
 ('FUJII Yuriko', datetime.date(1972, 12, 1), 'VENGE BALBOA Christian', datetime.date(1972, 12, 1)),
 ('PETRUNIV Vasyl', datetime.date(1973, 12, 31), 'SHIM Youngjip', datetime.date(1973, 12, 31)),
 ('ABU AWAD Khetam', datetime.date(1974, 10, 8), 'DETTONI Cristian', datetime.date(1974, 10, 8)),
 ('PENDER Dariusz', datetime.date(1974, 10, 16), 'di TORO Lisa Daniela', datetime.date(1974, 10, 16)),
 ('PALIKUCA Mitar', datetime.date(1974, 10, 21), 'SMETANINE David', datetime.date(1974, 10, 21)),
 ('JOHNSON-DWYER Philippa', datetime.date(1974, 12, 9), 'WILK Rafal', datetime.date(1974, 12, 9)),
 ('DAVID Krisztina', datetime.date(1975, 12, 5), 'DIXON Amy', datetime.date(1975, 12, 5)),
 ('CUNNINGTON Tammy', datetime.date(1975, 12, 27), 'VATNHAMAR Havard', datetime.date(1975, 12, 27)),
 ('KIM Kyung Hoon', datetime.date(1976, 3, 5), 'MALYSHEV Sergey', datetime.date(1976, 3, 5)),
 ('NOVAK Carol-Eduard', datetime.date(1976, 7, 28), 'de JESUS SANTOS Edneusa', datetime.date(1976, 7, 28)),
 ('KORVER Cher', datetime.date(1976, 8, 11), 'LIN Ya-Hsuan', datetime.date(1976, 8, 11)),
 ('MARVIG Henrik', datetime.date(1977, 3, 19), 'TOUPE David', datetime.date(1977, 3, 19)),
 ('BRUSSIG Carmen', datetime.date(1977, 5, 20), 'BRUSSIG Ramona', datetime.date(1977, 5, 20)),
 ('EBERHARDT Fernando', datetime.date(1977, 6, 21), 'JACOBS David', datetime.date(1977, 6, 21)),
 ('JEON Min Jae', datetime.date(1977, 7, 12), 'TATSUMI Hiromi', datetime.date(1977, 7, 12)),
 ('GUEDOUN Nabil', datetime.date(1977, 8, 4), 'SABINO Claudia Cicero', datetime.date(1977, 8, 4)),
 ('GUEDOUN Nabil', datetime.date(1977, 8, 4), 'de OLIVEIRA Jose Carlos Chagas', datetime.date(1977, 8, 4)),
 ('SABINO Claudia Cicero', datetime.date(1977, 8, 4), 'de OLIVEIRA Jose Carlos Chagas', datetime.date(1977, 8, 4)),
 ('BUCLAW Dorota', datetime.date(1977, 10, 16), 'EJIKE Lucy', datetime.date(1977, 10, 16)),
 ('DAME STOREY Sarah', datetime.date(1977, 10, 26), 'STOREY Sarah', datetime.date(1977, 10, 26)),
 ('JAKIMCZUK Tomasz', datetime.date(1978, 3, 17), 'SALEH Sameh Mohamed', datetime.date(1978, 3, 17)),
 ('AMADU Yusif', datetime.date(1978, 4, 6), 'AZEVEDO Adriana', datetime.date(1978, 4, 6)),
 ('BEJDIC Ervin', datetime.date(1978, 8, 10), 'de VRIES Tim', datetime.date(1978, 8, 10)),
 ('ANDRIEVSKAIA Tatiana', datetime.date(1978, 8, 15), 'LEEP Jose', datetime.date(1978, 8, 15)),
 ('KURODA Tomonari', datetime.date(1978, 10, 9), 'RISTIC Dragan', datetime.date(1978, 10, 9)),
 ('PATTAWAEO Phattharaphon', datetime.date(1978, 11, 6), 'SURANJI Laslo', datetime.date(1978, 11, 6)),
 ('BANG Yu-Jeong', datetime.date(1979, 2, 12), 'GOMEZ Maximiliano Ramon', datetime.date(1979, 2, 12)),
 ('LYSIAK Krystyna', datetime.date(1979, 3, 11), 'SALEHIHAJIKOLAEI Ramezan', datetime.date(1979, 3, 11)),
 ('ABOUELYAZEID Mohamed', datetime.date(1979, 3, 29), 'GOMES PESSANHA Michel', datetime.date(1979, 3, 29)),
 ('RODRIGUES Eric', datetime.date(1979, 5, 7), 'YANG Chao', datetime.date(1979, 5, 7)),
 ('MASSOUD Hossam Massoud', datetime.date(1979, 6, 13), 'MURATLI Nazmiye', datetime.date(1979, 6, 13)),
 ('BIABANI Ramezan', datetime.date(1979, 6, 25), 'ELELFAT Mohamed', datetime.date(1979, 6, 25)),
 ('BIABANI Ramezan', datetime.date(1979, 6, 25), 'STIBNERS Brett', datetime.date(1979, 6, 25)),
 ('ELELFAT Mohamed', datetime.date(1979, 6, 25), 'STIBNERS Brett', datetime.date(1979, 6, 25)),
 ('ASTANOV Elvin', datetime.date(1979, 7, 5), 'PEREIRA ARRUDA Harlley D

Write a query that generates a table like this...

country	    athlete_count	    medal_count	    medals_per_athlete

People's Republic of China	256	276	1.08

Netherlands	74	77	1.04

Italy	114	91	0.80

Ukraine	139	108	0.78

Great Britain	221	171	0.77

First I will try to get a list of unique countries for all athletes and order to check for duplicates.

In [42]:
%%sql
SELECT DISTINCT country FROM athletes ORDER BY country LIMIT 10

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
10 rows affected.


[('Afghanistan',),
 ('Algeria',),
 ('Angola',),
 ('Argentina',),
 ('Armenia',),
 ('Aruba',),
 ('Australia',),
 ('Austria',),
 ('Azerbaijan',),
 ('Bahrain',)]

Next I will try to get a table with a count of athletes per country.

In [43]:
%%sql
SELECT country, COUNT(*) AS athletes_per_country
FROM athletes
GROUP BY country
ORDER BY country
LIMIT 10

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
10 rows affected.


[('Afghanistan', 2),
 ('Algeria', 56),
 ('Angola', 2),
 ('Argentina', 59),
 ('Armenia', 1),
 ('Aruba', 1),
 ('Australia', 181),
 ('Austria', 25),
 ('Azerbaijan', 36),
 ('Bahrain', 2)]

I will make a table with count of medals per country.

In [44]:
%%sql
SELECT country, COUNT(*) AS medals_per_country
FROM medals
GROUP BY country
ORDER BY country
LIMIT 10

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
10 rows affected.


[('Algeria', 12),
 ('Argentina', 18),
 ('Australia', 101),
 ('Austria', 9),
 ('Azerbaijan', 19),
 ('Belarus', 7),
 ('Belgium', 15),
 ('Bosnia and Herzegovina', 11),
 ('Brazil', 111),
 ('Bulgaria', 2)]

I will Join these tables to show 3 columns.

In [45]:
%%sql 
SELECT table_a.country AS country, athletes_per_country, medals_per_country
FROM (
    SELECT country, COUNT(*) AS athletes_per_country
FROM athletes
GROUP BY country
) AS table_a
JOIN (
    SELECT country, COUNT(*) AS medals_per_country
FROM medals
GROUP BY country
) AS table_b
ON table_a.country = table_b.country
LIMIT 10


 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
10 rows affected.


[('Indonesia', 23, 11),
 ('Venezuela', 27, 7),
 ('Czech Republic', 29, 10),
 ('Sweden', 29, 9),
 ('Montenegro', 5, 1),
 ('Uganda', 4, 1),
 ('Jordan', 10, 5),
 ('Ireland', 31, 7),
 ('Singapore', 11, 2),
 ('Sri Lanka', 9, 2)]

Now I will add the final column and order.

In [51]:
%%sql 
SELECT table_a.country AS country, athlete_count, medal_count, round(
    CAST((CAST(medal_count AS float)/ athlete_count)AS numeric)
    , 2) AS medals_per_athlete
FROM (
    SELECT country, COUNT(*) AS athlete_count 
FROM athletes
GROUP BY country
) AS table_a
JOIN (
    SELECT country, COUNT(*) AS medal_count 
FROM medals
GROUP BY country
) AS table_b
ON table_a.country = table_b.country
ORDER BY medals_per_athlete DESC
LIMIT 5

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
5 rows affected.


[("People's Republic of China", 256, 276, Decimal('1.08')),
 ('Netherlands', 74, 77, Decimal('1.04')),
 ('Italy', 114, 91, Decimal('0.80')),
 ('Ukraine', 139, 108, Decimal('0.78')),
 ('Great Britain', 221, 171, Decimal('0.77'))]

In [55]:
%%sql 
SELECT 
    table_a.country AS country, athlete_count, medal_count, round(CAST(medal_count AS numeric)/ athlete_count, 2) AS medals_per_athlete
FROM (
    SELECT country, COUNT(*) AS athlete_count 
    FROM athletes
    GROUP BY country
    ) AS table_a
JOIN (
    SELECT country, COUNT(*) AS medal_count 
    FROM medals
    GROUP BY country
    ) AS table_b
ON 
    table_a.country = table_b.country
ORDER BY 
    medals_per_athlete DESC
LIMIT 5

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
5 rows affected.


[("People's Republic of China", 256, 276, Decimal('1.08')),
 ('Netherlands', 74, 77, Decimal('1.04')),
 ('Italy', 114, 91, Decimal('0.80')),
 ('Ukraine', 139, 108, Decimal('0.78')),
 ('Great Britain', 221, 171, Decimal('0.77'))]

Temu wanted to write a query that generates a table showing the number of athletes and gold medals for each country. Their code is below - it doesn't work as intended.

In [56]:
%%sql
SELECT medals_sub.country, medals_sub.medals, athletes_sub.athletes FROM (
    SELECT COUNT(*) as gold_medals, country
    FROM medals
    WHERE medal='Gold Medal'
    GROUP BY country
) medals_sub
INNER JOIN (
    SELECT COUNT(*) as athletes, countrycode FROM athletes GROUP BY countrycode
) athletes_sub
ON medals_sub.country = athletes_sub.countrycode;

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
(psycopg2.errors.UndefinedColumn) column medals_sub.medals does not exist
LINE 1: SELECT medals_sub.country, medals_sub.medals, athletes_sub.a...
                                   ^

[SQL: SELECT medals_sub.country, medals_sub.medals, athletes_sub.athletes FROM (
    SELECT COUNT(*) as gold_medals, country
    FROM medals
    WHERE medal='Gold Medal'
    GROUP BY country
) medals_sub
INNER JOIN (
    SELECT COUNT(*) as athletes, countrycode FROM athletes GROUP BY countrycode
) athletes_sub
ON medals_sub.country = athletes_sub.countrycode;]
(Background on this error at: http://sqlalche.me/e/14/f405)


In [61]:
%%sql
SELECT medals_sub.country, medals_sub.gold_medals, athletes_sub.athletes FROM (
    SELECT COUNT(*) as gold_medals, country
    FROM medals
    WHERE medal='Gold Medal'
    GROUP BY country
) medals_sub
INNER JOIN (
    SELECT COUNT(*) as athletes, country FROM athletes GROUP BY country
) athletes_sub
ON medals_sub.country = athletes_sub.country;

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
62 rows affected.


[('Indonesia', 4, 23),
 ('Venezuela', 3, 27),
 ('Czech Republic', 2, 29),
 ('Sweden', 1, 29),
 ('Jordan', 4, 10),
 ('Ireland', 4, 31),
 ('Singapore', 2, 11),
 ('Sri Lanka', 1, 9),
 ('Uzbekistan', 8, 44),
 ('Finland', 1, 16),
 ('Colombia', 3, 66),
 ('Ukraine', 24, 139),
 ('Cuba', 4, 17),
 ('Algeria', 4, 56),
 ('France', 13, 146),
 ('Slovakia', 7, 28),
 ('Israel', 6, 33),
 ('Malaysia', 3, 25),
 ('Cyprus', 1, 3),
 ('Turkey', 7, 87),
 ('Nigeria', 4, 22),
 ('Belarus', 5, 19),
 ('Netherlands', 38, 74),
 ('Islamic Republic of Iran', 23, 63),
 ('Australia', 24, 181),
 ('Serbia', 2, 20),
 ('Spain', 9, 137),
 ('United Arab Emirates', 1, 12),
 ('Belgium', 4, 34),
 ('Republic of Korea', 4, 85),
 ('Great Britain', 64, 221),
 ('Thailand', 8, 76),
 ("People's Republic of China", 127, 256),
 ('Italy', 19, 114),
 ('Germany', 13, 137),
 ('Canada', 5, 129),
 ('Azerbaijan', 14, 36),
 ('Greece', 1, 45),
 ('India', 5, 54),
 ('Chile', 2, 19),
 ('South Africa', 4, 34),
 ('Peru', 1, 11),
 ('Kazakhstan', 1, 26),
 ('Japan', 14, 262),
 ('Denmark', 3, 25),
 ('Mongolia', 1, 11),
 ('United States of America', 65, 243),
 ('Switzerland', 7, 21),
 ('Ecuador', 1, 8),
 ('New Zealand', 6, 32),
 ('Hungary', 7, 39),
 ('RPC', 49, 248),
 ('Norway', 2, 15),
 ('Pakistan', 1, 2),
 ('Brazil', 36, 241),
 ('Austria', 1, 25),
 ('Ethiopia', 1, 3),
 ('Tunisia', 4, 25),
 ('Morocco', 4, 36),
 ('Mexico', 7, 60),
 ('Poland', 8, 91),
 ('Costa Rica', 1, 9)]

Paula wanted to write a query that generates a table showing the average age of competitors winning each medal - gold, silver or bronze. Her query is below - does it work as intended? HINT: The answer is no :)

In [95]:
%%sql
SELECT subquery_result.medal, SUM(subquery_result.age) as average_age
FROM (
  SELECT
    medals.medal,
    DATE_PART('year', CURRENT_DATE) - DATE_PART('year', athletes.date_of_birth) as age
    FROM athletes INNER JOIN medals
    ON athletes.name = medals.name
) subquery_result
GROUP BY medal;


 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
3 rows affected.


[('Silver Medal', 23582.0), ('Gold Medal', 23247.0), ('Bronze Medal', 25988.0)]

The average age is much too high so there must be an issue calculating it. I notice looking at the code that the sum of age is being used rather than the average so I will change this first. 

In [96]:
%%sql
SELECT subquery_result.medal, AVG(subquery_result.age) as average_age
FROM (
  SELECT
    medals.medal,
    DATE_PART('year', CURRENT_DATE) - DATE_PART('year', athletes.date_of_birth) as age
    FROM athletes INNER JOIN medals
    ON athletes.name = medals.name
) subquery_result
GROUP BY medal;

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
3 rows affected.


[('Silver Medal', 32.79833101529903),
 ('Gold Medal', 31.88888888888889),
 ('Bronze Medal', 33.105732484076434)]

This appears to give much more sensible answers. However displaying these answers to such high precision is misleading as the calculated age for each athlete is an integer. Below I have corrected the precision to 0 decimal places. 

In [97]:
%%sql
SELECT subquery_result.medal, round(CAST(AVG(subquery_result.age) AS numeric), 0) as average_age
FROM (
  SELECT
    medals.medal,
    DATE_PART('year', CURRENT_DATE) - DATE_PART('year', athletes.date_of_birth) as age
    FROM athletes INNER JOIN medals
    ON athletes.name = medals.name
) subquery_result
GROUP BY medal;

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
3 rows affected.


[('Silver Medal', Decimal('33')),
 ('Gold Medal', Decimal('32')),
 ('Bronze Medal', Decimal('33'))]

This is more honest precision in terms of how the ages were calculated but could be better. 

My age is not the current year minus my birth year, the exact current and birth date matters. I can include months and days to make age a float rather than integer and be more accurate. 

I have displayed 5 decimal places below to compare to the years only values.

In [98]:
%%sql
SELECT subquery_result.medal, round(CAST(AVG(subquery_result.age) AS numeric), 5) as average_age
FROM (
  SELECT
    medals.medal,
    (DATE_PART('year', CURRENT_DATE) - DATE_PART('year', athletes.date_of_birth) 
     + ((DATE_PART('month', CURRENT_DATE) - DATE_PART('month', athletes.date_of_birth))/12)
     + ((DATE_PART('day', CURRENT_DATE) - DATE_PART('day', athletes.date_of_birth))/365.25))
    as age
    FROM athletes INNER JOIN medals
    ON athletes.name = medals.name
) subquery_result
GROUP BY medal;

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
3 rows affected.


[('Silver Medal', Decimal('32.77597')),
 ('Gold Medal', Decimal('31.85802')),
 ('Bronze Medal', Decimal('33.09113'))]

Below I round to their precision to 1 decimal place. 

In [99]:
%%sql
SELECT subquery_result.medal, round(CAST(AVG(subquery_result.age) AS numeric), 1) as average_age
FROM (
  SELECT
    medals.medal,
    (DATE_PART('year', CURRENT_DATE) - DATE_PART('year', athletes.date_of_birth) 
    + ((DATE_PART('month', CURRENT_DATE) - DATE_PART('month', athletes.date_of_birth))/12)
    + ((DATE_PART('day', CURRENT_DATE) - DATE_PART('day', athletes.date_of_birth))/365.25))
    as age
    FROM athletes INNER JOIN medals
    ON athletes.name = medals.name
) subquery_result
GROUP BY medal;

 * postgresql+psycopg2://localhost:5432/intro_to_sql_for_analytics
3 rows affected.


[('Silver Medal', Decimal('32.8')),
 ('Gold Medal', Decimal('31.9')),
 ('Bronze Medal', Decimal('33.1'))]